server/server.go

```diff
@@ -433,3 +444,66 @@ func ServeOnLocalTailscaled(ctx context.Context, lc *local.Client, st *ipnstate.
 
 	return func() { watcher.Close() }, watcherChan, nil
 }
+
+// writeHTTPError writes an appropriate HTTP error response based on the request's Accept header.
+// It logs the error with appropriate severity level and sends either JSON or plain text response.
+//
+// Parameters:
+//   - w: http.ResponseWriter to write the response
+//   - r: *http.Request to inspect Accept header and method/path for logging
+//   - statusCode: HTTP status code (e.g., 400, 401, 403, 500)
+//   - errorCode: Unique error code, use constants: ecAccessDenied, ecInvalidRequest, etc.
+//   - errorDescription: Human-readable error description
+//   - err: Optional underlying error for additional logging context
+func writeHTTPError(
+	w http.ResponseWriter,
+	r *http.Request,
+	statusCode int,
+	errorCode, errorDescription string,
+	err error,
+) {
+	args := []any{
+		slog.Int("status", statusCode),
+		slog.String("method", r.Method),
+		slog.String("path", r.URL.Path),
+		slog.String("error", errorCode),
```

Comment:
just to avoid conflict with the `err` argument added below
```suggestion
		slog.String("errorCode", errorCode),
```